core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1067,13 +1121,63 @@ class KafkaController(val config: KafkaConfig,
     (topicsToBeDeleted, topicsIneligibleForDeletion)
   }
 
-  private def updateLeaderAndIsrCache(partitions: Seq[TopicPartition] = controllerContext.allPartitions.toSeq): Unit = {
-    val leaderIsrAndControllerEpochs = zkClient.getTopicPartitionStates(partitions)
+  // sequential version:  very slow for clusters with many topic-partitions (e.g., > 50000)
+  private def updateLeaderAndIsrCache(partitions: Seq[TopicPartition]): Unit = {
+    val zkPartitionRecurseStartMs = time.milliseconds()
+    val leaderIsrAndControllerEpochs = zkClient.getTopicPartitionStates(partitions)  // 98.4% of method's time
+    info(s"partition-level init (sequential): finished recursing ${partitions.size} partitions in ZK ${timing(zkPartitionRecurseStartMs, -1)}")
     leaderIsrAndControllerEpochs.forKeyValue { (partition, leaderIsrAndControllerEpoch) =>
       controllerContext.putPartitionLeadershipInfo(partition, leaderIsrAndControllerEpoch)
     }
   }
 
+  // parallel version:  read topic-partitions info from ZK in multiple threads
+  private def updateLeaderAndIsrCacheParallel(partitions: Set[TopicPartition]): Unit = {
+    if (partitions.isEmpty)
+      return
+
+    val pool = Executors.newFixedThreadPool(config.liNumControllerInitThreads)
+    val leaderIsrAndControllerEpochs: mutable.Map[TopicPartition, LeaderIsrAndControllerEpoch] = mutable.Map.empty
+
+    // (1) split up partitions set into liNumControllerInitThreads batches
+    //     [not yet tested:  optionally could set numBatches = 3 * liNumControllerInitThreads (for example)
+    //     in order to limit occasional (rare) damage by a slow executor:  other N-1 executors that complete
+    //     in normal time will pick up additional small batches, thereby covering for slow one]
+    val numTopicPartitionsPerBatch = (partitions.size + config.liNumControllerInitThreads - 1) / config.liNumControllerInitThreads
+    val splitPartitions = partitions.grouped(numTopicPartitionsPerBatch)
+
+    // (2) loop over numThreads Seqs, wrap each in a Runnable that calls updateLeaderAndIsrCache() on it, and
+    //     submit Runnable to thread pool, storing all returned Futures in another Seq
+    val zkJobs = zkClients.zip(splitPartitions.toSeq).map { case (zkClientN, partitionsBatch) =>
```

Comment:
I think the arithmetic here is correct.
The doc of the zip method says if the two have different lengths, the extra ones in the longer collection will be ignored. If somehow that happens, it'll be quite hard to troubleshoot.
Can we throw an exception if their lengths don't match?

Reply:
Actually, shouldn't we just log a warning and fall back to the sequential version? Breaking the controller is pretty bad, at least outside of tests. As we've recently seen, rollback doesn't always work...